cloudinit: Enable cross compiling for cloudinit

prinzdezibel opened the pull request:
## This PR is part of a larger [PR](https://github.com/NixOS/nixpkgs/pull/363935). It was split to allow for easier reviewing. This part is about cross compiling cloud-init and network-manager.


## Testing/Reviewing
Example for cross compiling from aarch64 to x86_64:
```
nix-build '<nixpkgs>' -I nixpkgs=https://github.com/prinzdezibel/NixOS?ref=nixos-cross-compile-cloudinit    -A pkgsCross.gnu64.maturin
nix-build '<nixpkgs>' -I nixpkgs=https://github.com/prinzdezibel/NixOS?ref=nixos-cross-compile-cloudinit    -A pkgsCross.gnu64.openfortivpn
nix-build '<nixpkgs>' -I nixpkgs=https://github.com/prinzdezibel/NixOS?ref=nixos-cross-compile-cloudinit    -A pkgsCross.gnu64.networkmanager-fortisslvpn
nix-build '<nixpkgs>' -I nixpkgs=https://github.com/prinzdezibel/NixOS?ref=nixos-cross-compile-cloudinit    -A pkgsCross.gnu64.networkmanager-iodine
nix-build '<nixpkgs>' -I nixpkgs=https://github.com/prinzdezibel/NixOS?ref=nixos-cross-compile-cloudinit    -A pkgsCross.gnu64.cloud-init
nix-build '<nixpkgs>' -I nixpkgs=https://github.com/prinzdezibel/NixOS?ref=nixos-cross-compile-cloudinit    -A pkgsCross.gnu64.python312Packages.rpds-py

```


## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).


---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

prinzdezibel committed:
Enable full qemu emulation with TCG fallback when cross compiling images via make-disk-image

prinzdezibel committed:
Undo python fixes and move them to separate branch.

prinzdezibel committed:
--amend

prinzdezibel committed:
Enable dot file globbing

prinzdezibel committed:
Enable cross compiling of network manager plugin

prinzdezibel committed:
Revert "Undo python fixes and move them to separate branch."

This reverts commit 391d3afb00cabf2ab992434c9d1751c430ba6a74.

prinzdezibel committed:
Linting: Remove trailing whitespace.

prinzdezibel committed:
Remove wrong comment

prinzdezibel committed:
Update pkgs/by-name/ma/maturin/package.nix

Co-authored-by: seth <getchoo@tuta.io>

prinzdezibel committed:
Update nixos/lib/make-disk-image.nix

Co-authored-by: seth <getchoo@tuta.io>

prinzdezibel committed:
remove vm related stuff

prinzdezibel committed:
formatting

misuzu commented on pkgs/build-support/rust/hooks/default.nix:
```diff
@@ -86,6 +91,7 @@
       ];
       substitutions = {
         inherit (rust.envVars) rustTargetPlatformSpec setEnv;
+        pythonInterpreter = lib.versions.majorMinor pythonOnBuildForHost.version;
```
Pretty sure this is the wrong way since we build python packages for multiple python versions

misuzu commented on pkgs/by-name/ma/maturin/package.nix:
```diff
@@ -24,10 +24,12 @@ rustPlatform.buildRustPackage rec {
-    libiconv
-  ];
+  buildInputs =
+    [ python3 ]
```
It shouldn't depend on specific python version...

prinzdezibel replied:
Correct. This is not necessary.

misuzu commented on pkgs/development/python-modules/rpds-py/default.nix:
```diff
@@ -35,6 +36,7 @@ buildPythonPackage rec {
     rustPlatform.maturinBuildHook
     cargo
     rustc
+    python3
```
I don't think this is right, there should already be a specific python interpreter in `nativeBuildInputs`

prinzdezibel replied:
Correct. This is not needed.

misuzu commented:
I've opened four separate PRs for these changes to make things faster:

https://github.com/NixOS/nixpkgs/pull/365220
https://github.com/NixOS/nixpkgs/pull/365328
https://github.com/NixOS/nixpkgs/pull/365329
https://github.com/NixOS/nixpkgs/pull/365330

I've dropped maturin changes since they need another approach and not really needed to fix cloud-init.
I've also added `Co-authored-by` where I've reused your changes.

prinzdezibel commented:
@misuzu 
> I've dropped maturin changes since they need another approach and not really needed to fix cloud-init.
I've also added Co-authored-by where I've reused your changes.

To me it looks like the changes are indeed needed to fix cloud-init. I'm not able to cross compile cloud-init from master. The build fails during execution of maturin build hooks.

I've streamlined my changes to be aligned with the PR you mentioned (https://github.com/NixOS/nixpkgs/pull/279145).

Please find my new PR [here] (https://github.com/NixOS/nixpkgs/pull/365571).
Thanks.

misuzu commented:
> To me it looks like the changes are indeed needed to fix cloud-init. I'm not able to cross compile cloud-init from master. The build fails during execution of maturin build hooks.

The changes are in the staging branch.

prinzdezibel commented:
@misuzu : I don't think so?! 

Could you please have a look here:  https://github.com/NixOS/nixpkgs/pull/365571/files

Thanks!

misuzu commented:
> @misuzu : I don't think so?!

Check the staging branch, it's fixed there, not in master.
